src/librbd/ImageWatcher.cc

```diff
@@ -55,13 +55,19 @@ struct C_UnwatchAndFlush : public Context {
       r = rados.aio_watch_flush(aio_comp);
       assert(r == 0);
       aio_comp->release();
-    } else {
-      Context::complete(ret_val);
+      return;
     }
+
+    // ensure our reference to the RadosClient is released prior
+    // to completing the callback to avoid racing an explicit
+    // librados shutdown
+    Context *ctx = on_finish;
+    delete this;
+
+    ctx->complete(ret_val);
```

Comment:
@dillaman Ah, missed this: we can't use ret_val here. Need to copy it before delete.